handlers/common.go

```diff
@@ -130,27 +128,214 @@ func GetValidatorEarnings(validators []uint64, currency string) (*types.Validato
+	fiveMonths := time.Hour * 24 * 150
+
+	activeValidatorsCount := *services.GetLatestStats().ActiveValidatorCount
+	// Calculate the expected number of slot proposals for 30 days
+	expectedSlotProposals := calcExpectedSlotProposals(oneMonth, validatorsCount, activeValidatorsCount)
+
+	// Get the timeframe for which we should consider qualified proposals
+	var proposalTimeframe time.Duration
+	// Time since the first block in the proposed block slice
+	timeSinceFirstBlock := time.Since(utils.SlotToTime(slots[0]))
+
+	// Determine the appropriate timeframe based on the time since the first block and the expected slot proposals
+	switch {
+	case timeSinceFirstBlock < fiveDays:
+		proposalTimeframe = fiveDays
+	case timeSinceFirstBlock < oneWeek:
+		proposalTimeframe = oneWeek
+	case timeSinceFirstBlock < oneMonth:
+		proposalTimeframe = oneMonth
+	case timeSinceFirstBlock > fiveMonths && expectedSlotProposals <= 0.75:
+		proposalTimeframe = fiveMonths
+	case timeSinceFirstBlock > fourMonths && expectedSlotProposals <= 1:
+		proposalTimeframe = fourMonths
+	case timeSinceFirstBlock > threeMonths && expectedSlotProposals <= 1.4:
+		proposalTimeframe = threeMonths
+	case timeSinceFirstBlock > twoMonths && expectedSlotProposals <= 2.1:
+		proposalTimeframe = twoMonths
+	case timeSinceFirstBlock > sixWeeks && expectedSlotProposals <= 2.8:
+		proposalTimeframe = sixWeeks
+	default:
+		proposalTimeframe = oneMonth
+	}
+
+	// Recalculate expected slot proposals for the new timeframe
+	expectedSlotProposals = calcExpectedSlotProposals(proposalTimeframe, validatorsCount, activeValidatorsCount)
+	if expectedSlotProposals == 0 {
+		return 0
+	}
+	// Cutoff time for proposals to be considered qualified
+	blockProposalCutoffTime := time.Now().Add(-proposalTimeframe)
+
+	// Count the number of qualified proposals
+	qualifiedProposalCount := 0
+	for _, slot := range slots {
+		if utils.SlotToTime(slot).After(blockProposalCutoffTime) {
+			qualifiedProposalCount++
+		}
+	}
+	// Return the luck as the ratio of qualified proposals to expected slot proposals
+	return float64(qualifiedProposalCount) / expectedSlotProposals
+}
+
+// calcExpectedSlotProposals calculates the expected number of slot proposals for a certain time frame and validator count
+func calcExpectedSlotProposals(timeframe time.Duration, validatorCount int, activeValidatorsCount uint64) float64 {
+	if validatorCount == 0 || activeValidatorsCount == 0 {
+		return 0
+	}
+	slotsPerTimeframe := timeframe.Seconds() / float64(utils.Config.Chain.Config.SecondsPerSlot)
+	return (slotsPerTimeframe / float64(activeValidatorsCount)) * float64(validatorCount)
+}
+
+// getNextBlockEstimateTimestamp will return the estimated timestamp of the next block proposal
+// given the blocks proposed by the validators and the number of validators
+//
+// precondition: proposedBlocks is sorted by ascending block number
+func getNextBlockEstimateTimestamp(slots []uint64, validatorsCount int) *time.Time {
```

Comment:
- We only use the latest block that was proposed for the calculation so there is no need to give all of them as a parameter.
- Does it make sense to show the estimated time based on the that last block? If something would happen once a year based on some probability and he checks after (1 year - 5 seconds) the text would say `(exp. in 5s)` so the reader might expect there to be a really high chance for it to happen now but it is still as always a low "once a year" chance.
Maybe a text like `(exp. avg. once every 1 year)` would be more appropriate?

Reply:
after talking to @Buttaa we're going to show both in the tooltip, ptal